File: game-runtime/packages/sim/src/systems/loot-runtime.ts
import type { EquippedEquipmentInstance } from '@jwgb/content';
import { type EquipmentId, entityId, type Vec2Mm, vec2Mm } from '@jwgb/core';
import type { LootDrop, LootDropKind, MutableSimulationState, SimEvent } from '../types';

export const LOOT_SAFE_ZONE_EXPIRY = Number.MAX_SAFE_INTEGER;
export const LOOT_CURRENCY_EXPIRY_TICKS = 60 * 20;
export const LOOT_SHORT_EXPIRY_TICKS = 60 * 20;
export const LOOT_LONG_EXPIRY_TICKS = 120 * 20;
export const LOOT_STORM_GRACE_TICKS = 60 * 20;

export function createRuntimeLootDrop(
  state: MutableSimulationState,
  values: {
    readonly position: Vec2Mm;
    readonly gold?: number;
    readonly experience?: number;
    readonly gems?: number;
    readonly equipmentId?: EquipmentId | null;
    readonly bookPassiveId?: LootDrop['bookPassiveId'];
    readonly activeId?: LootDrop['activeId'];
    readonly kind: LootDropKind;
    readonly expiresAtTick?: number;
    readonly equipmentInstance?: EquippedEquipmentInstance | null;
  },
): LootDrop {
  const instance = values.equipmentInstance ?? null;
  const drop: LootDrop = {
    entityId: entityId(state.nextEntityId),
    position: vec2Mm(values.position.x, values.position.z),
    gold: values.gold ?? 0,
    experience: values.experience ?? 0,
    gems: values.gems ?? 0,
    equipmentId: values.equipmentId ?? null,
    bookPassiveId: values.bookPassiveId ?? null,
    createdAtTick: state.tick,
    expiresAtTick:
      values.expiresAtTick ??
      (values.kind === 'equipment' ||
      values.kind === 'death-equipment' ||
      values.kind === 'skill-book' ||
      values.kind === 'active'
        ? LOOT_SAFE_ZONE_EXPIRY
        : state.tick + LOOT_CURRENCY_EXPIRY_TICKS),
    kind: values.kind,
    activeId: values.activeId ?? null,
    equipmentInstanceId: instance?.instanceId ?? null,
    acquiredAtTick: instance?.acquiredAtTick ?? null,
    permanentAttackBonus: instance?.permanentAttackBonus ?? 0,
    stormCoveredSinceTick: null,
  };
  state.nextEntityId += 1;
  state.lootDrops.set(drop.entityId, drop);
  return drop;
}

export function emitLootDropped(
  state: MutableSimulationState,
  events: SimEvent[],
  drop: LootDrop,
  sourceEntityId: LootDrop['entityId'],
): void {
  events.push({
    type: 'loot-dropped',
    tick: state.tick,
    entityId: drop.entityId,
    sourceEntityId,
    gold: drop.gold,
    experience: drop.experience,
    gems: drop.gems,
    equipmentId: drop.equipmentId,
    bookPassiveId: drop.bookPassiveId,
    activeId: drop.activeId ?? null,
    ...(drop.kind === undefined ? {} : { lootKind: drop.kind }),
  });
}

export function createEquipmentLootDrop(
  state: MutableSimulationState,
  position: Vec2Mm,
  instance: EquippedEquipmentInstance,
  kind: 'equipment' | 'death-equipment' = 'equipment',
): LootDrop {
  return createRuntimeLootDrop(state, {
    position,
    equipmentId: instance.equipmentId,
    kind,
    equipmentInstance: instance,
    expiresAtTick: LOOT_SAFE_ZONE_EXPIRY,
  });
}

File: game-runtime/unity/Packages/com.jwgb.sim/Runtime/Deterministic/Systems/LootRuntime.cs
using System.Collections.Generic;
using Jwgb.Core;

namespace Jwgb.Sim.Deterministic
{
    /// <summary>
    /// Port of packages/sim/src/systems/loot-runtime.ts. All authoritative
    /// runtime loot producers go through here so the optional runtime keys
    /// (kind/activeId/equipmentInstanceId/acquiredAtTick/permanentAttackBonus/
    /// stormCoveredSinceTick) are always present in the state hash.
    /// </summary>
    internal static class LootRuntime
    {
        /// <summary>Mirrors Number.MAX_SAFE_INTEGER in the TS oracle.</summary>
        public const long SafeZoneExpiry = 9_007_199_254_740_991;

        public const int CurrencyExpiryTicks =
            60 * SimulationConstants.TicksPerSecond;

        public const int LongExpiryTicks =
            120 * SimulationConstants.TicksPerSecond;

        public const int StormGraceTicks =
            60 * SimulationConstants.TicksPerSecond;

        public static bool IsPersistentKind(string kind)
        {
            return kind == "equipment" ||
                kind == "death-equipment" ||
                kind == "skill-book" ||
                kind == "active";
        }

        public static LootDropState CreateRuntimeLootDrop(
            SimulationState state,
            Int2Mm position,
            string kind,
            int gold = 0,
            int experience = 0,
            int gems = 0,
            string equipmentId = null,
            string bookPassiveId = null,
            string activeId = null,
            long? expiresAtTick = null,
            EquippedEquipmentInstance? equipmentInstance = null)
        {
            var drop = new LootDropState
            {
                EntityId = state.NextEntityId,
                Position = position,
                Gold = gold,
                Experience = experience,
                Gems = gems,
                EquipmentId = equipmentId,
                BookPassiveId = bookPassiveId,
                CreatedAtTick = state.Tick,
                ExpiresAtTick = expiresAtTick ??
                    (IsPersistentKind(kind)
                        ? SafeZoneExpiry
                        : state.Tick + CurrencyExpiryTicks),
                HasRuntimeFields = true,
                Kind = kind,
                ActiveId = activeId,
                EquipmentInstanceId = equipmentInstance?.InstanceId,
                AcquiredAtTick = equipmentInstance?.AcquiredAtTick,
                PermanentAttackBonus =
                    equipmentInstance?.PermanentAttackBonus ?? 0,
                StormCoveredSinceTick = null
            };
            state.NextEntityId += 1;
            state.LootDrops.Add(drop.EntityId, drop);
            return drop;
        }

        public static LootDropState CreateEquipmentLootDrop(
            SimulationState state,
            Int2Mm position,
            EquippedEquipmentInstance instance,
            string kind = "equipment")
        {
            return CreateRuntimeLootDrop(
                state,
                position,
                kind,
                equipmentId: instance.EquipmentId,
                expiresAtTick: SafeZoneExpiry,
                equipmentInstance: instance);
        }

        public static void EmitLootDropped(
            SimulationState state,
            List<SimEvent> events,
            LootDropState drop,
            int sourceEntityId)
        {
            // Mirror the fixture event normalizer: loot-dropped keeps only
            // entityId, sourceEntityId and amount (= gold); every other field
            // must stay at its default so the harness comparison passes.
            events.Add(
                new SimEvent
                {
                    Type = "loot-dropped",
                    Tick = state.Tick,
                    EntityId = drop.EntityId,
                    SourceEntityId = sourceEntityId,
                    Amount = drop.Gold
                });
        }
    }
}
